### PR TITLE
[MM-37984] Allow Desktop App to authenticate via external providers outside of the app on supported servers

### PR DESCRIPTION
**Pull request by devinbinnie**
#### Summary
Based on https://github.com/mattermost/mattermost/pull/23795 which was reverted due to security issues.

This changes un-reverts the aforementioned PR, and adds a server-side generated token that will be passed to the Desktop App upon successful login via deep link.

This does change the flow a bit, meaning that the deep link will NEED to be clicked in order for the Desktop App to login.

This is definitely worth a complete re-review from UX, QA and Security to ensure everything is in place. Devs may also want to pay attention to the changes involving the server token.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-37984

#### Screenshots
See https://github.com/mattermost/mattermost/pull/23795

```release-note
Added API endpoint POST /api/v4/user/login/desktop_login
Modifield OAuth/SAML flows to include `desktop_login` where applicable
```

### CODE REVIEW

File: server/channels/db/migrations/mysql/000110_create_desktop_tokens.up.sql
Comment (agnivade) [inline server/channels/db/migrations/mysql/000110_create_desktop_tokens.up.sql:6]: The main fetch SQL query seems to use 

```
Where{sq.Eq{"DesktopToken": desktopToken},
sq.Eq{"ServerToken": serverToken},
sq.GtOrEq{"CreateAt": minCreateAt},
}
```

Do we want to have a compound index for that? Probably worth looking at which would be the most common queries and see if we want to add an index. Now would be the right time to do it, otherwise, it takes longer to create an index later.

File: server/channels/app/desktop_login.go
Comment (agnivade) [inline server/channels/app/desktop_login.go:26]: Any reason we are spawning so many goroutines unnecessarily in these methods? There are lots of pitfalls with unbounded goroutine creation, especially ones that make DB queries: https://developers.mattermost.com/contribute/more-info/server/style-guide/#default-to-sync-instead-of-async.

> Do not create one-off goroutines without knowing when/how they exit.

Unless there is a strong reason for it, the default behavior should be to _not_ use a goroutine.

File: server/channels/db/migrations/mysql/000110_create_desktop_tokens.up.sql
Comment (devinbinnie) [reply server/channels/db/migrations/mysql/000110_create_desktop_tokens.up.sql:6]: Yes, that's a good idea. We originally had an index on just the PK and createat which was enough, I'll change it to a compound of the 3.

File: server/channels/app/desktop_login.go
Comment (devinbinnie) [reply server/channels/app/desktop_login.go:26]: So I did this as a `defer` originally because I didn't want to lock the database by deleting a row synchronously. If this is not a major concern, I will remove the goroutine and go back to doing it synchronously.

Comment (agnivade) [reply server/channels/app/desktop_login.go:26]: The locking will anyways happen regardless of you use a goroutine or not. It's just that now the load on the DB won't be unbounded.

File: server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql
Comment (agnivade) [inline server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql:9]: `CREATE INDEX CONCURRENTLY` for postgres: https://developers.mattermost.com/contribute/more-info/server/schema-migration-guide/#details

**Review by agnivade — CHANGES_REQUESTED** [review]

File: server/channels/app/desktop_login.go
Comment (larkox) [inline server/channels/app/desktop_login.go:56]: We are deleting in case of any error. Shouldn't we just return in `GetUserId` any token with that client token, and do the rest of the checks here? That way, we only try to delete if the token has to be deleted. That way we will have oftentimes one less database call.

File: server/channels/store/sqlstore/desktop_tokens_store.go
Comment (larkox) [inline server/channels/store/sqlstore/desktop_tokens_store.go:59]: Probably not a big deal, but `desktopToken` is a string provided through the API. Any need to validate it? squirrel should deal with any weird characters and I guess we are just making comparisons with this string, so there shouldn't be any issue, but I wanted to raise it just in case.

File: webapp/channels/src/components/desktop_auth_token.tsx
Comment (larkox) [inline webapp/channels/src/components/desktop_auth_token.tsx:103]: Why do you need the casting?

File: webapp/channels/src/components/desktop_auth_token.tsx
Comment (larkox) [inline webapp/channels/src/components/desktop_auth_token.tsx:52]: This is a boolean, not a React.ReactNode, and probably should default to false.

**Review by larkox — APPROVED** [review]
LGTM, just some nits and questions.

File: server/channels/store/sqlstore/desktop_tokens_store.go
Comment (esarafianou) [reply server/channels/store/sqlstore/desktop_tokens_store.go:59]: Yep Squirrel produces a [prepared statement](https://en.wikipedia.org/wiki/Prepared_statement) and the RDBMS takes care of combining the query with the inputs.

File: server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql
Comment (devinbinnie) [reply server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql:9]:
```suggestion
CREATE INDEX CONCURRENTLY IF NOT EXISTS idx_desktoptokens_desktoptoken_servertoken_createat ON desktoptokens(desktoptoken, servertoken, createat)
```

File: server/channels/api4/user.go
Comment (devinbinnie) [reply server/channels/api4/user.go:1975]: It adds that time negative to the time requested. So if it's 3:03PM it becomes 3:00PM
We do this to check if the created at date is after this time so that we know if the token is expired or not.

File: server/channels/app/desktop_login.go
Comment (devinbinnie) [reply server/channels/app/desktop_login.go:68]: Correct, we don't want to support that case because then we'd likely have to record some kind of "session ID" to keep track of both and it gets more complicated.

It's also here for security to make sure that someone isn't going to try and steal a session while the token is being sent back after validation.

File: server/channels/app/desktop_login.go
Comment (devinbinnie) [reply server/channels/app/desktop_login.go:56]: Depending on the error, we can't necessarily confirm whether the token was deleted or not (eg. a DB lock or communication error of some kind, cosmic rays and such :P). I'd like to be very persistent about making sure that it is deleted no matter what to force the user to restart the flow. Keeps things clean IMO.

File: webapp/channels/src/components/desktop_auth_token.tsx
Comment (devinbinnie) [reply webapp/channels/src/components/desktop_auth_token.tsx:52]: Whoops, why'd I do that...

**Review by abhijit-singh — COMMENTED** [review]
Thanks @devinbinnie but it doesn't look it's working as expected on the test server. I'm unable to login on the desktop app even after successful authentication on the browser. Attaching a screen-cap. 

I'm assuming this does not need a new build of the desktop app but maybe I'm missing something?

https://github.com/mattermost/mattermost/assets/14369607/033558a7-c5f8-4d4c-927e-e97a63e4b21f

File: server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql
Comment (agnivade) [inline server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql:9]: semicolon missing at the end

**Review by agnivade — APPROVED** [review]
Just one minor thing. Good to go after that. Pre-approving to expediate.

File: server/channels/api4/user.go
Comment (larkox) [reply server/channels/api4/user.go:1975]: I wonder if this is responsibility of the caller or responsibility of the app layer, but 0/5

File: server/channels/web/oauth.go
Comment (esarafianou) [inline server/channels/web/oauth.go:400]: We don't need to save the desktop token on the server side.  The desktop token will be generated in the desktop, passed as a query parameter to the external idP and then returned to the desktop app via deeplink once the user has been redirected to the landing page.  It's the desktop that needs to store the desktop token so that it can verify it matches the original copy once it receives it again.

File: server/channels/web/oauth.go
Comment (esarafianou) [inline server/channels/web/oauth.go:478]: Same comment, we don't need to save the desktop token on the server side.

File: server/channels/web/saml.go
Comment (esarafianou) [inline server/channels/web/saml.go:65]: Same comment, we don't need to save the desktop token on the server side.

File: server/channels/web/oauth.go
Comment (esarafianou) [inline server/channels/web/oauth.go:356]: This is not needed since we don't need to store the client desktop token on the server side at all.

File: server/channels/app/desktop_login.go
Comment (esarafianou) [inline server/channels/app/desktop_login.go:35]: The server_token is stored next to the authN user ID along with a TTL, not next to the desktop token.

File: server/channels/web/oauth.go
Comment (esarafianou) [inline server/channels/web/oauth.go:382]: Here we need to append the desktop client token as a query parameter too, so that the desktop can verify it against its local copy.

File: server/channels/web/saml.go
Comment (esarafianou) [inline server/channels/web/saml.go:202]: Same comment as above in the oauth, this is not needed because we don't need to store the desktop token at all

File: server/channels/web/saml.go
Comment (esarafianou) [inline server/channels/web/saml.go:224]: Same comment as above, here we need to append the desktop client token as a query parameter too, so that the desktop can verify it against its local copy.

**Review by esarafianou — CHANGES_REQUESTED** [review]
@devinbinnie let me know if you'd like to discuss again the flow with the 2 tokens.

File: server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql
Comment (devinbinnie) [reply server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql:9]:
```suggestion
CREATE INDEX CONCURRENTLY IF NOT EXISTS idx_desktoptokens_desktoptoken_servertoken_createat ON desktoptokens(desktoptoken, servertoken, createat);
```

File: server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql
Comment (devinbinnie) [reply server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql]: @agnivade Adding this causes tests to fail. Going to remove for now, is my syntax wrong? It looks like it's supposed to work: https://www.postgresql.org/docs/current/sql-createindex.html

**Comment by devinbinnie** [comment]
Re-requesting reviews across the board as the logic has changed.

File: server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql
Comment (agnivade) [inline server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql:1]: What will happen to this migration? Once deployed, this table has already been created. We would need to sync up with the SRE team to actually drop the table, and let it be re-created again.

File: server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql
Comment (agnivade) [inline server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql:8]:
```suggestion
CREATE INDEX CONCURRENTLY IF NOT EXISTS idx_desktoptokens_token_createat ON desktoptokens(token, createat)
```

**Review by agnivade — CHANGES_REQUESTED** [review]

File: server/channels/web/saml.go
Comment (larkox) [inline server/channels/web/saml.go:217]: This piece of code looks almost the same to the one on the oauth file. Any chance we can somehow reuse it and write it just once?

**Review by larkox — APPROVED** [review]
Just one small nit

**Review by esarafianou — APPROVED** [review]
LGTM. I'd just like a clarification on one comment that I left

**Review by abhijit-singh — CHANGES_REQUESTED** [review]
@devinbinnie Looks like I'm getting the same behaviour as last time.  The desktop app is still stuck on the redirecting to browser screen even after logging in on the browser and clicking on `Open Mattermost`

File: server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql
Comment (devinbinnie) [reply server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql:1]: I guess the easiest thing to do would be to just keep the original migration in place, then do an ALTER to make the changes. I'll do that.

File: server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql
Comment (devinbinnie) [reply server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql]: Getting this error with `CONCURRENTLY`:

`Error: failed to initialize platform: cannot create store: failed to apply database migrations: driver: postgres, message: failed to execute migration, command: executing_query, originalError: pq: CREATE INDEX CONCURRENTLY cannot run inside a transaction block`

File: webapp/channels/src/components/desktop_auth_token.tsx
Comment (devinbinnie) [reply webapp/channels/src/components/desktop_auth_token.tsx:64]: I think it's possible but potentially unexpected. I had this logic here before when we were polling the server, but I guess that's not needed anymore.

File: server/channels/web/saml.go
Comment (devinbinnie) [reply server/channels/web/saml.go:217]: We probably could, but I think refactoring it out would probably cause some confusion and wouldn't really save many lines. 0/5 tbh though.

**Comment by devinbinnie** [comment]
@abhijit-singh Found an issue for certain sites not updating when you deep link. Should be working now :) tested on the current SpinWick

**Review by abhijit-singh — APPROVED** [review]
Looks good, thanks @devinbinnie!

File: server/channels/db/migrations/migrations.list
Comment (agnivade) [inline server/channels/db/migrations/migrations.list:222]: I see that you are using 112 separately. There is no need to add 110 anymore. The migration from 110 has already been deployed and exists on cloud.

File: server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql
Comment (agnivade) [reply server/channels/db/migrations/postgres/000110_create_desktop_tokens.up.sql]: Oops yeah my bad. You need to move this statement to a separate file and add a `// morph:nontransactional` comment on top. But since this is just creating an index on a new table, it should be fine to skip it.

File: server/channels/db/migrations/migrations.list
Comment (devinbinnie) [reply server/channels/db/migrations/migrations.list:222]: Just adding it back since this PR (including the migration) was reverted. So if it's not added here it won't be in the codebase at all.

**Review by yasserfaraazkhan — APPROVED** [review]
Thank you @devinbinnie 
was able to run the desktop app locally. 
clicking on Google took use to browser and open the desktop after successful login.

Screenshot 2023-08-29 at 10 11 36 PM

Screenshot 2023-08-29 at 10 11 39 PM

File: server/channels/db/migrations/migrations.list
Comment (agnivade) [reply server/channels/db/migrations/migrations.list:222]: Right, that's fine.

**Review by agnivade — APPROVED** [review]
Nicely done :)

**Comment by cwarnermm** [comment]
@devinbinnie - Are there product documentation updates needed for this PR?

**Comment by devinbinnie** [comment]
> @devinbinnie - Are there product documentation updates needed for this PR?

@cwarnermm Yes, we might want to inform users that the login process will take them outside of the Desktop App if they are using an external provider. Do we have a good spot to document that?

**Comment by cwarnermm** [comment]
@devinbinnie - Yes. I've assigned https://github.com/mattermost/docs/pull/6700 to you as a reviewer.

**Pull request merged**
